### PR TITLE
EP/MIN 311 Pool total supply/borrow and TVL

### PR DESCRIPTION
**Pull request by Apollo50**
**Description:**

Fix RPC protocolTotalValues.
Add test.
Update types.json

**Changes in storage or API:**

Describe storage structure changes or API changes. Make sure to update the [API docs](https://minterestfinance.atlassian.net/wiki/spaces/MINTEREST/pages/134807557/Minterest+API).

**Checklist**

- [x] Code is covered with tests
- [ ] Code is covered with benchmarks
- [ ] Fresh benchmark results included

### CODE REVIEW

File: pallets/controller/rpc/runtime-api/src/lib.rs [hunk 1]
```diff
@@ -53,6 +53,20 @@ pub struct BalanceInfo {
 	pub amount: Balance,
 }
 
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Eq, PartialEq, Encode, Decode, Default, RuntimeDebug)]
+pub struct BalanceInfoExtended {
+	#[cfg_attr(feature = "std", serde(serialize_with = "serialize_as_string"))]
+	#[cfg_attr(feature = "std", serde(deserialize_with = "deserialize_from_string"))]
+	pub pool_total_supply: Balance,
```
Comment (lut1k) [inline pallets/controller/rpc/runtime-api/src/lib.rs:61]:
```suggestion
	pub pools_total_supply_usd: Balance,
```
or
```suggestion
	pub pools_total_supply_in_usd: Balance,
```
Same for `pool_total_borrow`

File: pallets/controller/rpc/runtime-api/src/lib.rs [hunk 2]
```diff
@@ -53,6 +53,20 @@ pub struct BalanceInfo {
 	pub amount: Balance,
 }
 
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Eq, PartialEq, Encode, Decode, Default, RuntimeDebug)]
+pub struct BalanceInfoExtended {
```
Comment (lut1k) [inline pallets/controller/rpc/runtime-api/src/lib.rs:58]:
```suggestion
pub struct ProtocolTotalValue {
```

File: pallets/controller/src/lib.rs [hunk 3]
```diff
@@ -577,38 +577,89 @@ impl<T: Config> Pallet<T> {
...
-					.ok_or(Error::<T>::BalanceOverflow)?;
-				Ok(current_value
-					.checked_add(pool_total_value_usd)
-					.ok_or(Error::<T>::BalanceOverflow)?)
-			})?;
-		Ok(total_value)
+			.try_fold(
+				(Balance::zero(), Balance::zero(), Balance::zero()),
+				|current_value, &pool_id| -> result::Result<(Balance, Balance, Balance), DispatchError> {
+					let (protocol_available_liquidity, protocol_borrow, tvl) = current_value;
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:602]:
```suggestion
				|(protocol_available_liquidity, protocol_borrow, tvl), &pool_id| -> result::Result<(Balance, Balance, Balance), DispatchError> {
```

File: pallets/controller/src/lib.rs [hunk 4]
```diff
@@ -577,38 +577,89 @@ impl<T: Config> Pallet<T> {
 		Ok((total_supply_balance, total_borrowed_balance))
 	}
 
-	/// Calculates total amount of money currently held in the protocol in usd.
-	/// Total value is calculated as: sum(total_issuance_n * exchange_rate_n * oracle_price_n),
+	/// Calculates pool_total_supply, pool_total_borrow including interest and excluding
+	/// protocol_interest, tvl (Total Value Locked) in USD
```
Comment (lut1k) [inline pallets/controller/src/lib.rs:581]: Why do we exclude `protocol_interest` ?

**Review by lut1k — APPROVED** [review]

File: pallets/controller/rpc/runtime-api/src/lib.rs [hunk 5]
```diff
@@ -53,6 +53,23 @@ pub struct BalanceInfo {
...
+	pub pool_total_supply_in_usd: Balance,
+	#[cfg_attr(feature = "std", serde(serialize_with = "serialize_as_string"))]
+	#[cfg_attr(feature = "std", serde(deserialize_with = "deserialize_from_string"))]
+	pub pool_total_borrow_in_usd: Balance,
+	#[cfg_attr(feature = "std", serde(serialize_with = "serialize_as_string"))]
+	#[cfg_attr(feature = "std", serde(deserialize_with = "deserialize_from_string"))]
+	pub tvl_in_usd: Balance,
+	#[cfg_attr(feature = "std", serde(serialize_with = "serialize_as_string"))]
+	#[cfg_attr(feature = "std", serde(deserialize_with = "deserialize_from_string"))]
+	pub pool_total_interest_in_usd: Balance,
```
[the remark this answers is not in the document]
Comment (Apollo50) [reply pallets/controller/rpc/runtime-api/src/lib.rs:70]:
```suggestion
	pub pool_total_protocol_interest_in_usd: Balance,
```
?

**Review by lut1k — APPROVED** [review]

**Pull request merged**